nixos/amdvlk: init module

JohnRTitor opened the pull request:
## Description of changes

- Adds `hardware.amdgpu.amdvlk` option.
- This module enables amdvlk via `hardware.opengl.extraPackages`.
- The main use case I see for this module is configuring [runtime settings](https://github.com/GPUOpen-Drivers/AMDVLK?tab=readme-ov-file#runtime-settings) of AMDVLK.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

CC: @Flakebi maintainer of AMDVLK

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

JohnRTitor committed:
nixos/amvlk: init module

JohnRTitor committed:
nixos/amdvlk: add settings option

JohnRTitor commented:
```nix
  hardware.amdgpu.amdvlk.enable = true;
  hardware.amdgpu.amdvlk.settings = {
    AllowVkPipelineCachingToDisk = 1;
    ShaderCacheMode = 1;
    IFH = 0;
    EnableVmAlwaysValid = 1;
    IdleAfterSubmitGpuMask = 1;
  };
```

![image](https://github.com/NixOS/nixpkgs/assets/50095635/b5fd3a11-4a73-4045-8fc0-40892adae6da)

JohnRTitor commented:
Okay, should be ready for review.

SuperSandro2000 commented on nixos/modules/services/hardware/amdvlk.nix:
```diff
@@ -0,0 +1,65 @@
+  };
+
+  config = lib.mkIf cfg.enable {
+    hardware.opengl = lib.mkMerge [
+      {
+        enable = lib.mkDefault true;
+        driSupport = lib.mkDefault true;
+        extraPackages = [ cfg.package ];
+      }
+      (lib.mkIf cfg.support32Bit.enable {
+        driSupport32Bit = lib.mkDefault true;
+        extraPackages32 = [ cfg.support32Bit.package ];
+      })
+    ];
```
```suggestion
    hardware.opengl = {
      enable = true;
      driSupport = true;
      extraPackages = [ cfg.package ];
      driSupport32Bit = lib.mkIf cfg.support32Bit.enable true;
      extraPackages32 = [ cfg.support32Bit.package ];
    }
```
extraPackages32 is completely unused, if driSupport32Bit is false.
Also we don't need to use mkDefault because we want to enable those options if the module is enabled and people should know, if they are disabled

JohnRTitor replied:
Done in https://github.com/NixOS/nixpkgs/pull/318175/commits/52e48f6f78cd1c3c5216caac8e58471b2303f213

JohnRTitor committed:
nixos/amdvlk: cleanup

remove mkDefault: Users should know if 32Bit config is disabled and enable it accordingly
settings: type should be either integer or string

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4062

pbsds commented on nixos/modules/services/hardware/amdvlk.nix:
```diff
@@ -0,0 +1,61 @@
+    support32Bit.package = lib.mkPackageOption pkgs.driversi686Linux "amdvlk" { };
+
+    settings = lib.mkOption {
+      type = lib.types.attrs;
```
```suggestion
      type = lib.types.attrsOf lib.types.int;
```

JohnRTitor replied:
Sometimes, the values may contain a string, in particular for `GpuProfilerConfig.LogDirectory` it would be a directory path. See https://github.com/GPUOpen-Drivers/AMDVLK?tab=readme-ov-file#pal-gpuprofiler-layer

pbsds replied:
```suggestion
      type = with lib.types; attrsOf (either str int);
```

Yes! Please add a note in the 24.11 release notes :)

JohnRTitor replied:
Done in https://github.com/NixOS/nixpkgs/pull/318175/commits/85814898f2105be9960a2e234e5f16ba8d8ac265

Added 24.11 release notes in https://github.com/NixOS/nixpkgs/pull/318175/commits/b2f1ce3c111364629444363dc325bfbf6ca2cb65

pbsds reviewed (commented):
i built and checked the nixos docs

JohnRTitor committed:
docs/release-notes: 24.11: add amdvlk module

JohnRTitor commented:
@ofborg eval

pbsds commented on nixos/modules/services/hardware/amdvlk.nix:
```diff
@@ -0,0 +1,61 @@
+    supportExperimental.enable = lib.mkEnableOption "Experimental features support";
+
+    support32Bit.enable = lib.mkEnableOption "32-bit driver support";
+    support32Bit.package = lib.mkPackageOption pkgs.driversi686Linux "amdvlk" { };
```
to fix 
![image](https://github.com/NixOS/nixpkgs/assets/140964/18ffcd02-f263-4413-95f1-fec81ffbd069)


```suggestion
    support32Bit.package = lib.mkPackageOption pkgs [ "driversi686Linux" "amdvlk" ] { };
```

JohnRTitor replied:
Fixed in https://github.com/NixOS/nixpkgs/pull/318175/commits/56123e0d5b87a743df1aa82e5ad462cfbaaaa1b7

JohnRTitor commented:
This should be fine for the most part. By the way, @pbsds how are you building the nixos manual? I am asking because I could not find any related docs online.

`nix-build doc` just builds the simple and compact manual.

pbsds commented:
I found it in the [nixpkgs flake](https://github.com/NixOS/nixpkgs/blob/c870e19d09a9f539fd322039843d1cf705c65f3c/flake.nix#L74) and i combine it with `refs/pull/<id>/merge`, which is like `refs/pull/<id>/head` but merged on top of master

```
nix build github:nixos/nixpkgs/refs/pull/318175/merge#htmlDocs.nixosManual.x86_64-linux
```

Flakebi commented:
The nixos-hardware change reminded me that we have a bit of documentation for amdvlk in `nixos/doc/manual/configuration/gpu-accel.chapter.md`. Although it’s not wrong with this change, it can probably be updated to use the new options added here.

Mic92 commented:
Can we backport this? Required in https://github.com/NixOS/nixos-hardware/pull/977

pbsds commented:
On it, i'll do a manual one that does not touch the release notes

pbsds commented:
please view https://github.com/NixOS/nixpkgs/pull/319510 and https://github.com/NixOS/nixpkgs/pull/319512

JohnRTitor commented:
I don't think we'll need a backport to 23.11, it'll go EOL soon. 

For backport to 24.05, it should be added to 24.05 release notes.

pbsds commented:
I at least still run two AMD machines on 2311 which are too busy running builds for me to upgrade them.

I was of the impression that the release notes are frozen after release. Please correct me if I am wrong.

JohnRTitor commented:
> I was of the impression that the release notes are frozen after release. Please correct me if I am wrong.

It's in the PR template, not sure if that's documented elsewhere

![Screenshot_20240613-160655_Chrome](https://github.com/NixOS/nixpkgs/assets/50095635/446c4fab-71d1-439d-916f-11fb469ba751)

Mic92 commented:
> I at least still run two amdgpu machines with nixos-hardware/common-amd-gpu on 2311 which are too busy running builds for me to upgrade them.
> 
> I was of the impression that the release notes are frozen after release. Please correct me if I am wrong.

For new features, that are not breaking existing behavior, we can make exceptions.

pbsds commented:
I added the release notes.

> It's in the PR template, not sure if that's documented elsewhere

Thanks, i did not notice that change

---

This is a breaking change for all amdgpu consumers of nixos-hardware, since they need to adapt their config to append `.enable`. After thinking about it for a while I'm skewing towards renaming the the `hardware.amdgpu.amdvlk` option in this PR to avoid the collision or even reverting it. I'll open a third PR to that end and let you decide which ones to merge and close.

<sup>This would've been way simple if nixos-hardware was versioned like home-manager is.</sup>

caffineehacker commented:
Could this have been gated on the system.stateVersion instead of always being active? That would've avoided breaking current users and would allow for nixos-hardware to detect the stateVersion and decide what settings to use.

JohnRTitor commented:
As a followup to this, I have opened #319865, hopefully that the end of this.

Please check that out!
